src/coreclr/jit/morph.cpp

```diff
@@ -10243,6 +10243,18 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
                 {
                     op1 = fgMorphRetInd(tree->AsUnOp());
                 }
+                // Local assertions may enable zero or copy prop.
+                //
+                if (fgGlobalMorph && optLocalAssertionProp && (optAssertionCount > 0) && op1->OperIs(GT_LCL_VAR))
+                {
+                    GenTree* newOp1 = op1;
+                    while (newOp1 != nullptr)
+                    {
+                        op1    = newOp1;
+                        newOp1 = optAssertionProp(apFull, newOp1, nullptr, nullptr);
```

Comment:
It would feel like a lesser evil to "retry" replacing the local with its promoted field in post-order to me. It would be more in line with how the other decomposition transforms work.

Reply:
I agree it seems like a more natural fit. 

I had a version that did that, but it caused LSRA churn (though not too much).

Reply:
> I had a version that did that, but it caused LSRA churn (though not too much).

It seems a bit surprising it did -- wouldn't the lack of this replacement cause the failure (assert) being fixed?

To be clear, I meant trying the replacement on both pre- and post- order.

Reply:
Yes, I was trying it both places. I'll revive that version....

Reply:
Looks like my earlier assessment was wrong; doing the update both pre and post order works nicely and has no diffs on win x64 or x86. 

Updated.